src/hazelcore/dynamics/Evaluator.re

```diff
@@ -4,23 +4,27 @@ type ground_cases =
   | Ground
   | NotGroundOrHole(HTyp.t) /* the argument is the corresponding ground type */;
 
-let grounded_Arrow = NotGroundOrHole(Arrow(Hole, Hole));
-let grounded_Sum = NotGroundOrHole(Sum(Hole, Hole));
+let grounded_Arrow =
+  NotGroundOrHole(Arrow(Unknown(Internal), Unknown(Internal)));
+let grounded_Sum =
+  NotGroundOrHole(Sum(Unknown(Internal), Unknown(Internal)));
 let grounded_Prod = length =>
-  NotGroundOrHole(Prod(ListUtil.replicate(length, HTyp.Hole)));
-let grounded_List = NotGroundOrHole(List(Hole));
+  NotGroundOrHole(
+    Prod(ListUtil.replicate(length, HTyp.Unknown(Internal))),
+  );
+let grounded_List = NotGroundOrHole(List(Unknown(Internal)));
 
 let ground_cases_of = (ty: HTyp.t): ground_cases =>
   switch (ty) {
-  | Hole => Hole
+  | Unknown(_) => Hole
   | Bool
   | Int
   | Float
-  | Arrow(Hole, Hole)
-  | Sum(Hole, Hole)
-  | List(Hole) => Ground
+  | Arrow(Unknown(_), Unknown(_))
+  | Sum(Unknown(_), Unknown(_))
+  | List(Unknown(_)) => Ground
   | Prod(tys) =>
-    if (List.for_all(HTyp.eq(HTyp.Hole), tys)) {
+    if (List.for_all(HTyp.eq(HTyp.Unknown(Internal)), tys)) {
```

Comment:
I don't think this file should be restricted to Internal unknowns, TypHole unknowns are semantically the same thing at this point. need to use wildcards / replace the eq calls with pattern matching...